### ISSUE TITLE
Use a github token for checksums?

### ISSUE THREAD

**Issue by till** [cloudalchemy/ansible-node-exporter#165]
**What is missing?**

I think the role should support the use of a Github Token to avoid rate limiting.

Currently, I keep running into a 400 (Bad Request) errors when I deploy lots of nodes in parallel. That seems to rate-limiting from Github's end. In my environment, I cd to about 20 customer setups, each of them may have multiple nodes. The part we are continuously deploying involves a basic monitoring setup on each node (including the node-exporter).

**Why do we need it?**

We run Ansible in a container, shared-nothing so to speak. The container gets invoked during CI (merge to main branch).

There is no caching between builds, it seems to work well. There are no side effects, except for this part where each run makes requests against Github (API) resources and seems to run into the rate-limit eventually.

I was digging around, it seems that both of these blocks happen every time (unless I download the binary myself):

Checksum gathering:
https://github.com/cloudalchemy/ansible-node-exporter/blob/b9cb0eeccfaa104af5ba23e80c4104fef6814d93/tasks/preflight.yml#L98-L110

Downloading:
https://github.com/cloudalchemy/ansible-node-exporter/blob/c6ffcfdef305c39273e7ab13bea69374c92eb171/tasks/install.yml#L20-L52

I would think I wouldn't download anything — unless I really needed? Do you have any thoughts on changing that?

**Environment**

* Role version:

    `0.21.3`

* Ansible version information:

```
root@5c649b77871a:/ansible-all-the-things# ansible --version
ansible 2.9.8
  config file = /ansible-all-the-things/ansible.cfg
  configured module search path = ['/ansible-all-the-things/library']
  ansible python module location = /usr/local/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.7.8 (default, Jun 30 2020, 18:36:05) [GCC 8.3.0]
```

**Anything else we need to know?**:

**Comment by till** [comment]
OK, sorry about that — just checked main.yml — at least the downloading is not happening every time.

But the checksum fetching is.

**Comment by paulfantom** [comment]
Checksum fetching doesn't require github token as this is not accessing github api: https://github.com/cloudalchemy/ansible-node-exporter/blob/master/tasks/preflight.yml#L101. So having token won't change anything. It might be good to add some retries though.

**Comment by till** [comment]
I am almost sure everything does. And anything downloads is an implicit API. Any idea how to dump the complete response from a 400 to see all headers?

**Comment by SuperQ** [comment]
The real root cause here is hitting github URLs for every node, rather than using `delegate_to: localhost`.

If we add `delegate_to: localhost` to the checksum fetching, it will eliminate the rate limiting issue.

**Comment by till** [comment]
@SuperQ @paulfantom Can we re-open this? Because it doesn't fix the issue:

```
# curl   -H "Accept: application/vnd.github.v3+json"   https://api.github.com/rate_limit| jq '.resources.core'
{
  "limit": 60,
  "remaining": 60,
  "reset": 1595273758
}
```

I am burning through rate-limits in no time.

**Comment by till** [comment]
Putting it on `localhost` now emphasises this problem on a single server. My PR at least contained a `when` to not do the checksum lookup unless necessary.

**Comment by paulfantom** [comment]
Just for future reference. 

Role is using GitHub API **only** for checking latest available version of node_exporter. Other operations involving GitHub are not executed via `api.github.com`, but `github.com` and are not restricted by GitHub rate limiter (at least not the one which can be omitted with authenticated requests).

Rate limitation in GitHub API results in 403 not 400 error code response ([source](https://developer.github.com/v3/#rate-limiting)). 400 responses can be caused due to various issues in user environment (DNS misconfiguration, proxy settings, etc.)

After https://github.com/cloudalchemy/ansible-node-exporter/pull/169 was merged, the role executes downloading checksum only once on localhost. This is independent from the number of hosts in inventory and it is a bare minimum for checksum validation to work.

**Comment by till** [comment]
This is the log with `ANSIBLE_DEBUG`:

```
TASK [cloudalchemy.node-exporter : Get checksum list from github] **************
task path: /root/.ansible/roles/cloudalchemy.node-exporter/tasks/preflight.yml:116
    14 1595332896.31838: sending task start callback
    14 1595332896.31843: entering _queue_task() for hostname.etc.pp/set_fact
   238 1595332896.46968: running TaskExecutor() for hostname.etc.pp/TASK: cloudalchemy.node-exporter : Get checksum list from github
    14 1595332896.44633: worker is 1 (out of 1 available)
    14 1595332896.48459: exiting _queue_task() for hostname.etc.pp/set_fact
    14 1595332896.48470: done queuing things up, now waiting for results queue to drain
    14 1595332896.48476: waiting for pending results...
   238 1595332896.51284: in run() - task 0242c0a8-400f-f0cd-df0e-000000000054
   238 1595332896.52143: calling self._execute()
   238 1595332896.57482: Loading FilterModule 'core' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/core.py (found_in_cache=True, class_only=False)
   238 1595332896.57502: Loading FilterModule 'gcp_kms_filters' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/gcp_kms_filters.py (found_in_cache=True, class_only=False)
   238 1595332896.57515: Loading FilterModule 'ipaddr' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/ipaddr.py (found_in_cache=True, class_only=False)
   238 1595332896.57526: Loading FilterModule 'json_query' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/json_query.py (found_in_cache=True, class_only=False)
   238 1595332896.57538: Loading FilterModule 'k8s' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/k8s.py (found_in_cache=True, class_only=False)
   238 1595332896.57547: Loading FilterModule 'mathstuff' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/mathstuff.py (found_in_cache=True, class_only=False)
   238 1595332896.57555: Loading FilterModule 'network' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/network.py (found_in_cache=True, class_only=False)
   238 1595332896.57563: Loading FilterModule 'urls' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/urls.py (found_in_cache=True, class_only=False)
   238 1595332896.57570: Loading FilterModule 'urlsplit' from /usr/local/lib/python3.7/site-packages/ansible/plugins/filter/urlsplit.py (found_in_cache=True, class_only=False)
   238 1595332896.57772: Loading TestModule 'core' from /usr/local/lib/python3.7/site-packages/ansible/plugins/test/core.py (found_in_cache=True, class_only=False)
   238 1595332896.58437: Loading TestModule 'files' from /usr/local/lib/python3.7/site-packages/ansible/plugins/test/files.py (found_in_cache=True, class_only=False)
   238 1595332896.58548: Loading TestModule 'mathstuff' from /usr/local/lib/python3.7/site-packages/ansible/plugins/test/mathstuff.py (found_in_cache=True, class_only=False)
   238 1595332896.83223: Loaded config def from plugin (lookup/url)
   238 1595332896.83243: Loading LookupModule 'url' from /usr/local/lib/python3.7/site-packages/ansible/plugins/lookup/url.py
url lookup connecting to https://github.com/prometheus/node_exporter/releases/download/v1.0.1/sha256sums.txt
   238 1595332897.61232: done running TaskExecutor() for hostname.etc.pp/TASK: cloudalchemy.node-exporter : Get checksum list from github [0242c0a8-400f-f0cd-df0e-000000000054]
   238 1595332897.61267: sending task result for task 0242c0a8-400f-f0cd-df0e-000000000054
   238 1595332897.61321: done sending task result for task 0242c0a8-400f-f0cd-df0e-000000000054
   238 1595332897.61329: WORKER PROCESS EXITING
    14 1595332897.65028: marking hostname.etc.pp as failed
    14 1595332897.65049: getting the next task for host hostname.etc.pp
    14 1595332897.65063: done getting next task for host hostname.etc.pp
    14 1595332897.65072:  ^ task is: TASK: cloudalchemy.node-exporter : Get checksum for {{ go_arch }} architecture
    14 1595332897.65080:  ^ state is: HOST STATE: block=3, task=5, rescue=0, always=0, run_state=ITERATING_TASKS, fail_state=FAILED_NONE, pending_setup=False, tasks child state? (HOST STATE: block=0, task=5, rescue=0, always=0, run_state=ITERATING_TASKS, fail_state=FAILED_NONE, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
    14 1595332897.65090: marking host hostname.etc.pp failed, current state: HOST STATE: block=3, task=5, rescue=0, always=0, run_state=ITERATING_TASKS, fail_state=FAILED_NONE, pending_setup=False, tasks child state? (HOST STATE: block=0, task=4, rescue=0, always=0, run_state=ITERATING_TASKS, fail_state=FAILED_NONE, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
    14 1595332897.65100: ^ failed state is now: HOST STATE: block=3, task=5, rescue=0, always=0, run_state=ITERATING_TASKS, fail_state=FAILED_NONE, pending_setup=False, tasks child state? (HOST STATE: block=0, task=4, rescue=0, always=0, run_state=ITERATING_COMPLETE, fail_state=FAILED_TASKS, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
    14 1595332897.65106: getting the next task for host hostname.etc.pp
    14 1595332897.65115: done getting next task for host hostname.etc.pp
    14 1595332897.65120:  ^ task is: None
    14 1595332897.65125:  ^ state is: HOST STATE: block=3, task=5, rescue=0, always=0, run_state=ITERATING_COMPLETE, fail_state=FAILED_TASKS, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
    14 1595332897.65130: getting the next task for host hostname.etc.pp
    14 1595332897.65136: done getting next task for host hostname.etc.pp
    14 1595332897.65141:  ^ task is: None
    14 1595332897.65145:  ^ state is: HOST STATE: block=3, task=5, rescue=0, always=0, run_state=ITERATING_COMPLETE, fail_state=FAILED_TASKS, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
fatal: [hostname.etc.pp]: FAILED! => {
    "msg": "An unhandled exception occurred while running the lookup plugin 'url'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Received HTTP error for https://github.com/prometheus/node_exporter/releases/download/v1.0.1/sha256sums.txt : HTTP Error 400: Bad Request"
}
    14 1595332897.65203: no more pending results, returning what we have
    14 1595332897.65234: results queue empty
    14 1595332897.65239: checking for any_errors_fatal
    14 1595332897.65248: getting the next task for host hostname.etc.pp
    14 1595332897.65253: done getting next task for host hostname.etc.pp
    14 1595332897.65256:  ^ task is: None
    14 1595332897.65260:  ^ state is: HOST STATE: block=3, task=5, rescue=0, always=0, run_state=ITERATING_COMPLETE, fail_state=FAILED_TASKS, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
    14 1595332897.65263: done checking for any_errors_fatal
    14 1595332897.65265: checking for max_fail_percentage
    14 1595332897.65268: done checking for max_fail_percentage
    14 1595332897.65271: checking to see if all hosts have failed and the running result is not ok
    14 1595332897.65273: ^ not ok, so returning result now
```

**Comment by paulfantom** [comment]
We had [a similar issue in ansible-prometheus](https://github.com/cloudalchemy/ansible-prometheus/issues/138). It seems to me that url lookup plugin in ansible might be broken and it may be beneficial to go back to a combination of `get_url` and `set_fact` tasks.

**Comment by paulfantom** [comment]
Or someone would need to debug https://github.com/ansible/ansible/blame/devel/lib/ansible/plugins/lookup/url.py plugin :shrug:

**Comment by till** [comment]
I'll dig around. Can you re-open this, or do I need to make another ticket?

**Comment by till** [comment]
Replacing `lookup()` with `uri`:

```
    - name: Get checksum list from github
      uri:
        url: "https://github.com/prometheus/node_exporter/releases/download/v{{ node_exporter_version }}/sha256sums.txt"
        method: GET
        return_content: true
      register: _checksum_result
      until: _checksum_result.status == 200
      retries: 5

    - name: Set _checksums
      set_fact:
        _checksums: "{{ _checksum_result.stdout_lines }}"
      run_once: true
```

Yields the `400` — but not sure where `Authorization` header is introduced?

```
<localhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
--
2888 | <localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir /root/.ansible/tmp/ansible-tmp-1595343177.2519362-233-56170830947126 && echo ansible-tmp-1595343177.2519362-233-56170830947126="` echo /root/.ansible/tmp/ansible-tmp-1595343177.2519362-233-56170830947126 `" ) && sleep 0'
2889 | Using module file /usr/local/lib/python3.7/site-packages/ansible/modules/net_tools/basics/uri.py
2890 | Pipelining is enabled.
2891 | <localhost> EXEC /bin/sh -c '/usr/local/bin/python && sleep 0'
2892 | <localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1595343177.2519362-233-56170830947126/ > /dev/null 2>&1 && sleep 0'
2893 | FAILED - RETRYING: Get checksum list from github (1 retries left).Result was: {
2894 | "attempts": 5,
2895 | "changed": false,
2896 | "connection": "close",
2897 | "content": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidArgument</Code><Message>Only one auth mechanism allowed; only the X-Amz-Algorithm query parameter, Signature query string parameter or the Authorization header should be specified</Message><ArgumentName>Authorization</ArgumentName><ArgumentValue>Basic **redacted**</ArgumentValue><RequestId>C27DCB3881334C01</RequestId><HostId>FIqOQQeYf2vmrxT1CEeevSvccLgS7KYSLaduFK8FjCsbkHoUDMtcQfE5RLzPE8HWKvjJIj9ozzA=</HostId></Error>",
2898 | "content_type": "application/xml",
2899 | "date": "Tue, 21 Jul 2020 14:52:58 GMT",
2900 | "elapsed": 0,
2901 | "invocation": {
2902 | "module_args": {
2903 | "attributes": null,
2904 | "backup": null,
2905 | "body": null,
2906 | "body_format": "raw",
2907 | "client_cert": null,
2908 | "client_key": null,
2909 | "content": null,
2910 | "creates": null,
2911 | "delimiter": null,
2912 | "dest": null,
2913 | "directory_mode": null,
2914 | "follow": false,
2915 | "follow_redirects": "safe",
2916 | "force": false,
2917 | "force_basic_auth": false,
2918 | "group": null,
2919 | "headers": {},
2920 | "http_agent": "ansible-httpget",
2921 | "method": "GET",
2922 | "mode": null,
2923 | "owner": null,
2924 | "regexp": null,
2925 | "remote_src": null,
2926 | "removes": null,
2927 | "return_content": true,
2928 | "selevel": null,
2929 | "serole": null,
2930 | "setype": null,
2931 | "seuser": null,
2932 | "src": null,
2933 | "status_code": [
2934 | 200
2935 | ],
2936 | "timeout": 30,
2937 | "unix_socket": null,
2938 | "unsafe_writes": null,
2939 | "url": "https://github.com/prometheus/node_exporter/releases/download/v1.0.1/sha256sums.txt",
2940 | "url_password": null,
2941 | "url_username": null,
2942 | "use_proxy": true,
2943 | "validate_certs": true
2944 | }
2945 | },
2946 | "msg": "Status code was 400 and not [200]: HTTP Error 400: Bad Request",
2947 | "redirected": false,
2948 | "retries": 6,
2949 | "server": "AmazonS3",
2950 | "status": 400,
2951 | "transfer_encoding": "chunked",
2952 | "url": "https://github.com/prometheus/node_exporter/releases/download/v1.0.1/sha256sums.txt",
2953 | "x_amz_id_2": "FIqOQQeYf2vmrxT1CEeevSvccLgS7KYSLaduFK8FjCsbkHoUDMtcQfE5RLzPE8HWKvjJIj9ozzA=",
2954 | "x_amz_request_id": "C27DCB3881334C01"
2955 | }
```
